client-core/src/wallet/syncer.rs

```diff
@@ -277,14 +294,74 @@ impl<'a, S: SecureStorage, C: Client, D: TxDecryptor, F: FnMut(ProgressReport) -
         Ok(())
     }
 
+    fn do_recover_address(&mut self, transaction: &Transaction) -> Result<()> {
+        let _transaction_id = transaction.id();
+
+        let mut refetch = false;
+
+        let _inputs = transaction.inputs().to_vec();
+
+        let outputs = transaction.outputs().to_vec();
+
+        for (_i, output) in outputs.iter().enumerate() {
+            let newaddress = output.address.to_string();
+            let tmp_refetch = self.env.wallet_client.check_address(
+                &newaddress,
+                &self.env.name,
+                &self.env.enckey,
+            )?;
+            if tmp_refetch {
+                refetch = true;
+            }
+        }
+
+        if refetch {
+            log::info!("refetch address ---------------------------");
+            let wallet = service::load_wallet(&self.env.storage, &self.env.name, &self.env.enckey)
+                .expect("get wallet");
+            self.wallet = wallet.expect("get wallet");
+        }
+        Ok(())
+    }
+
+    fn recover_address(&mut self, blocks: &[FilteredBlock]) -> Result<()> {
+        let enclave_txids = blocks
+            .iter()
+            .flat_map(|block| block.enclave_transaction_ids.iter().copied())
+            .collect::<Vec<_>>();
+        let enclave_txs = self.env.decryptor.decrypt_tx(&enclave_txids)?;
+        let enclave_transactions = enclave_txs
+            .iter()
+            .map(|tx| (tx.id(), tx))
+            .collect::<HashMap<_, _>>();
+
+        for block in blocks {
+            for txid in block.enclave_transaction_ids.iter() {
+                if let (Some(tx), Some(_fee)) = (
+                    enclave_transactions.get(txid),
+                    block.valid_transaction_fees.get(txid),
+                ) {
+                    self.do_recover_address(&tx)?;
+                }
+            }
+        }
+
+        Ok(())
+    }
+
     fn handle_batch(&mut self, blocks: NonEmpty<FilteredBlock>) -> Result<()> {
         let enclave_txids = blocks
             .iter()
             .flat_map(|block| block.enclave_transaction_ids.iter().copied())
             .collect::<Vec<_>>();
         let enclave_txs = self.env.decryptor.decrypt_tx(&enclave_txids)?;
 
-        let memento = handle_blocks(&self.wallet, &self.wallet_state, &blocks, &enclave_txs)
+        if crate::types::WalletKind::HD == self.wallet.wallet_kind {
+            // only hdwallet
+            self.recover_address(&blocks)?;
```

Comment:
Even with HD wallet, should we make this feature optional with a command line option?

Reply:
ok, i'll add option for client-rpc

Reply:
> ok, i'll add option for client-rpc

Does client-cli support this feature?

Reply:
client-cli,too.
I'll add option

Reply:
yes, i'll add option